File: 06-binary-search-trees/20250801-beginner/go/solution_test.go
package main

import (
	"reflect"
	"testing"
)

func TestBSTBasicOperations(t *testing.T) {
	tests := []struct {
		name        string
		operations  []interface{}
		expected    []interface{}
	}{
		{
			name: "Test 1: Insert single value and search",
			operations: []interface{}{
				[]string{"insert", "search"},
				[]int{5, 5},
			},
			expected: []interface{}{true},
		},
		{
			name: "Test 2: Insert multiple values and search existing",
			operations: []interface{}{
				[]string{"insert", "insert", "insert", "search", "search", "search"},
				[]int{10, 5, 15, 5, 10, 15},
			},
			expected: []interface{}{true, true, true},
		},
		{
			name: "Test 3: Search for non-existing value",
			operations: []interface{}{
				[]string{"insert", "insert", "search"},
				[]int{10, 20, 15},
			},
			expected: []interface{}{false},
		},
		{
			name: "Test 4: In-order traversal of simple tree",
			operations: []interface{}{
				[]string{"insert", "insert", "insert", "traversal"},
				[]int{10, 5, 15, 0},
			},
			expected: []interface{}{[]int{5, 10, 15}},
		},
		{
			name: "Test 5: In-order traversal with more complex tree",
			operations: []interface{}{
				[]string{"insert", "insert", "insert", "insert", "insert", "traversal"},
				[]int{50, 30, 20, 40, 70, 0},
			},
			expected: []interface{}{[]int{20, 30, 40, 50, 70}},
		},
		{
			name: "Test 6: Insert duplicate values (should not create duplicates)",
			operations: []interface{}{
				[]string{"insert", "insert", "insert", "traversal"},
				[]int{10, 10, 10, 0},
			},
			expected: []interface{}{[]int{10}},
		},
		{
			name: "Test 7: Left-skewed tree operations",
			operations: []interface{}{
				[]string{"insert", "insert", "insert", "search", "traversal"},
				[]int{30, 20, 10, 20, 0},
			},
			expected: []interface{}{true, []int{10, 20, 30}},
		},
		{
			name: "Test 8: Right-skewed tree operations",
			operations: []interface{}{
				[]string{"insert", "insert", "insert", "search", "traversal"},
				[]int{10, 20, 30, 25, 0},
			},
			expected: []interface{}{false, []int{10, 20, 30}},
		},
		{
			name: "Test 9: Empty tree operations",
			operations: []interface{}{
				[]string{"search", "traversal"},
				[]int{10, 0},
			},
			expected: []interface{}{false, []int{}},
		},
		{
			name: "Test 10: Large set insertion and search",
			operations: []interface{}{
				[]string{"insert", "insert", "insert", "insert", "insert", "insert", "insert", "search", "search", "traversal"},
				[]int{50, 25, 75, 12, 37, 62, 87, 37, 100, 0},
			},
			expected: []interface{}{true, false, []int{12, 25, 37, 50, 62, 75, 87}},
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			bst := NewBST()
			operations := tt.operations[0].([]string)
			values := tt.operations[1].([]int)
			expectedResults := tt.expected
			resultIndex := 0

			for i, op := range operations {
				switch op {
				case "insert":
					bst.Insert(values[i])
				case "search":
					result := bst.Search(values[i])
					if result != expectedResults[resultIndex] {
						t.Errorf("Search operation failed. Expected %v, got %v", expectedResults[resultIndex], result)
					}
					resultIndex++
				case "traversal":
					result := bst.InOrderTraversal()
					expectedSlice := expectedResults[resultIndex].([]int)
					if !reflect.DeepEqual(result, expectedSlice) {
						t.Errorf("Traversal operation failed. Expected %v, got %v", expectedSlice, result)
					}
					resultIndex++
				}
			}
		})
	}
}

func TestBSTEdgeCases(t *testing.T) {
	t.Run("Test nil BST operations", func(t *testing.T) {
		bst := NewBST()
		if bst == nil {
			t.Skip("BST constructor returns nil - implement NewBST() first")
		}
		
		// Test operations on empty tree
		if bst.Search(10) != false {
			t.Error("Search on empty tree should return false")
		}
		
		traversal := bst.InOrderTraversal()
		if traversal == nil {
			traversal = []int{}
		}
		if len(traversal) != 0 {
			t.Error("Traversal on empty tree should return empty slice")
		}
	})
}

File: 06-binary-search-trees/20250801-beginner/ts/solution.test.ts
import { BST } from './solution';

describe('BST Basic Operations', () => {
    
    test('Test 1: Insert single value and search', () => {
        const bst = new BST();
        bst.insert(5);
        expect(bst.search(5)).toBe(true);
    });

    test('Test 2: Insert multiple values and search existing', () => {
        const bst = new BST();
        bst.insert(10);
        bst.insert(5);
        bst.insert(15);
        
        expect(bst.search(5)).toBe(true);
        expect(bst.search(10)).toBe(true);
        expect(bst.search(15)).toBe(true);
    });

    test('Test 3: Search for non-existing value', () => {
        const bst = new BST();
        bst.insert(10);
        bst.insert(20);
        
        expect(bst.search(15)).toBe(false);
    });

    test('Test 4: In-order traversal of simple tree', () => {
        const bst = new BST();
        bst.insert(10);
        bst.insert(5);
        bst.insert(15);
        
        expect(bst.inOrderTraversal()).toEqual([5, 10, 15]);
    });

    test('Test 5: In-order traversal with more complex tree', () => {
        const bst = new BST();
        const values = [50, 30, 20, 40, 70];
        values.forEach(val => bst.insert(val));
        
        expect(bst.inOrderTraversal()).toEqual([20, 30, 40, 50, 70]);
    });

    test('Test 6: Insert duplicate values (should not create duplicates)', () => {
        const bst = new BST();
        bst.insert(10);
        bst.insert(10);
        bst.insert(10);
        
        expect(bst.inOrderTraversal()).toEqual([10]);
    });

    test('Test 7: Left-skewed tree operations', () => {
        const bst = new BST();
        bst.insert(30);
        bst.insert(20);
        bst.insert(10);
        
        expect(bst.search(20)).toBe(true);
        expect(bst.inOrderTraversal()).toEqual([10, 20, 30]);
    });

    test('Test 8: Right-skewed tree operations', () => {
        const bst = new BST();
        bst.insert(10);
        bst.insert(20);
        bst.insert(30);
        
        expect(bst.search(25)).toBe(false);
        expect(bst.inOrderTraversal()).toEqual([10, 20, 30]);
    });

    test('Test 9: Empty tree operations', () => {
        const bst = new BST();
        
        expect(bst.search(10)).toBe(false);
        expect(bst.inOrderTraversal()).toEqual([]);
    });

    test('Test 10: Large set insertion and search', () => {
        const bst = new BST();
        const values = [50, 25, 75, 12, 37, 62, 87];
        values.forEach(val => bst.insert(val));
        
        expect(bst.search(37)).toBe(true);
        expect(bst.search(100)).toBe(false);
        expect(bst.inOrderTraversal()).toEqual([12, 25, 37, 50, 62, 75, 87]);
    });
});

describe('BST Edge Cases', () => {
    test('Empty tree search and traversal', () => {
        const bst = new BST();
        
        expect(bst.search(42)).toBe(false);
        expect(bst.inOrderTraversal()).toEqual([]);
    });

    test('Single node operations', () => {
        const bst = new BST();
        bst.insert(42);
        
        expect(bst.search(42)).toBe(true);
        expect(bst.search(10)).toBe(false);
        expect(bst.inOrderTraversal()).toEqual([42]);
    });
});